[Style] Convert the @page size descriptor to strong style types

weinig opened the pull request:
#### 3a95b4b3a45ce4cb1515b4355131456218593cec
<pre>
[Style] Convert the @page size descriptor to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=298480">https://bugs.webkit.org/show_bug.cgi?id=298480</a>

Reviewed by Darin Adler and Tim Nguyen.

Converts the @page `size` descriptor to use strong style types.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/page/PrintContext.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilder.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/page/StylePageSize.cpp: Added.
* Source/WebCore/style/values/page/StylePageSize.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:

Canonical link: <a href="https://commits.webkit.org/299726@main">https://commits.webkit.org/299726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb37ac767b73cc841ff8f67b782b506ab68121ea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119955 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39648 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30299 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126284 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72028 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121831 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40343 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48225 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91095 "") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60403 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122907 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32221 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107555 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71651 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31251 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25662 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69921 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101698 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25850 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129201 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46875 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35537 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99711 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47241 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103745 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99556 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25283 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45008 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23021 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43496 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46737 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52443 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46203 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49552 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47889 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3bb15f5f4babdb3f3c094f9f328d10a1b95b8794. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50395

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3352,38 +3352,31 @@ void Document::pageSizeAndMarginsInPixels(int pageIndex, IntSize& pageSize, int&
-    }
-    pageSize = IntSize(width, height);
+    pageSize = WTF::switchOn(style->pageSize(),
+        [&](const CSS::Keyword::Auto&) -> IntSize {
+            return pageSize;
+        },
+        [&](const CSS::Keyword::Landscape&) -> IntSize {
+            if (pageSize.width() < pageSize.height())
+                return pageSize.transposedSize();
+            return pageSize;
+        },
+        [&](const CSS::Keyword::Portrait&) -> IntSize {
+            if (pageSize.width() > pageSize.height())
+                return pageSize.transposedSize();
+            return pageSize;
+        },
```
```suggestion
        [&](const CSS::Keyword::Auto&) {
            return pageSize;
        },
        [&](const CSS::Keyword::Landscape&) {
            if (pageSize.width() < pageSize.height())
                return pageSize.transposedSize();
            return pageSize;
        },
        [&](const CSS::Keyword::Portrait&) {
            if (pageSize.width() > pageSize.height())
                return pageSize.transposedSize();
            return pageSize;
        },
```

darinadler commented on Source/WebCore/page/PrintContext.cpp:
```diff
@@ -399,8 +399,22 @@ String PrintContext::pageProperty(LocalFrame* frame, const String& propertyName,
+            [&](const CSS::Keyword::Portrait&) -> String {
+                return "portrait"_s;
+            },
+            [&](const Style::PageSize::Lengths& lengths) -> String {
+                return makeString(lengths.width().value, ' ', lengths.height().value);
+            }
```
```suggestion
            [&](const Style::PageSize::Lengths& lengths) {
                return makeString(lengths.width().value, ' ', lengths.height().value);
            }
```

darinadler commented on Source/WebCore/style/values/page/StylePageSize.cpp:
```diff
@@ -0,0 +1,188 @@
+        return Length<CSS::Nonnegative>(CSS::pixelsPerInch * inch);
+    };
+
+    static constexpr Length<CSS::Nonnegative> a5Width(mmLength(148));
```
Do these all end up getting compile time computed and constant folded?

weinig replied:
Yep.

darinadler commented on Source/WebCore/style/values/page/StylePageSize.h:
```diff
@@ -0,0 +1,60 @@
+
+// <'size'> (for @page) = <length [0,∞]>{1,2} | auto | [ <page-size> || [ portrait | landscape ] ]
+// https://drafts.csswg.org/css-page-3/#descdef-page-size
+struct PageSize {
```
Seems a bit unconventional to make this a `struct` but have the data member be private.

weinig replied:
Yeah, but it's been nice to have all the style types be struct for easier forward declaration.

weinig committed:
[Style] Convert the @page size descriptor to strong style types
https://bugs.webkit.org/show_bug.cgi?id=298480

Reviewed by Darin Adler and Tim Nguyen.

Converts the @page `size` descriptor to use strong style types.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/page/PrintContext.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilder.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/page/StylePageSize.cpp: Added.
* Source/WebCore/style/values/page/StylePageSize.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:

Canonical link: https://commits.webkit.org/299726@main
